hal/src/common/dmac/channel.rs

```diff
@@ -0,0 +1,313 @@
+
+#[cfg(any(
+    feature = "samd51",
+    feature = "same51",
+    feature = "same53",
+    feature = "same54"
+))]
+use crate::target_device::dmac::CHANNEL;
+
+use crate::target_device::DMAC;
+use crate::typelevel::Sealed;
+
+/// Uninitialized channel
+pub struct Uninitialized;
+/// Initialized and ready to transfer channel
+pub struct Ready;
+/// Busy channel
+pub struct Busy;
+
+/// DMA channel, capable of executing
+/// [`DmaTransfer`](super::transfer::DmaTransfer)s.
+pub struct Channel<Status, const ID: u8> {
+    _status: Status,
+}
+
+pub(crate) fn new_chan<const ID: u8>() -> Channel<Uninitialized, ID> {
+    Channel {
+        _status: Uninitialized,
+    }
+}
+
+impl<S, const ID: u8> Sealed for Channel<S, ID> {}
+
+/// These methods may be used on any DMA channel in any configuration
+impl<S, const ID: u8> Channel<S, ID> {
+    /// Set channel ID and run the closure. A closure is needed to ensure
+    /// the registers are accessed in an interrupt-safe way, as the SAMD21
+    /// DMAC is a little funky - It requires setting the channel number in
+    /// the CHID register, then access the channel control registers.
+    /// If an interrupt were to change the CHID register, we would be faced
+    /// with undefined behaviour.
+    #[cfg(any(feature = "samd11", feature = "samd21"))]
+    fn with_chid<F: Fn(&mut DMAC)>(&mut self, dmac: &mut DMAC, fun: F) {
+        cortex_m::interrupt::free(|_| {
+            // SAFETY: this is actually safe as long as we write a correct channel number to
+            // the CHID register
+            unsafe {
+                dmac.chid.modify(|_, w| w.id().bits(ID));
+            }
+
+            fun(dmac);
+        });
+    }
+
+    /// Set channel ID and run the closure. A closure is needed to ensure
+    /// the registers are accessed in an interrupt-safe way, as the SAMD21
+    /// DMAC is a little funky. For the SAMD51/SAMEx, we simply take a reference
+    /// to the correct channel number and run the closure on that.
+    #[cfg(any(
+        feature = "samd51",
+        feature = "same51",
+        feature = "same53",
+        feature = "same54"
+    ))]
+    fn with_chid<F: Fn(&CHANNEL)>(&mut self, dmac: &mut DMAC, fun: F) {
+        let mut ch = &dmac.channel[ID as usize];
+        fun(&mut ch);
+    }
+
+    /// Configure the DMA channel so that it is ready to be used by a
+    /// [`DmaTransfer`](super::transfer::DmaTransfer).
+    ///
+    /// # Return
+    ///
+    /// A `Channel` with a `Ready` status
+    pub fn init(
+        mut self,
+        lvl: PriorityLevel,
+        enable_interrupts: bool,
+        controller: &mut DmaController,
+    ) -> Channel<Ready, ID> {
+        let dmac = controller.as_mut();
+
+        // Software reset the channel for good measure
+        self._reset_private(dmac);
+
+        self.with_chid(dmac, |d| {
+            // Reset the channel to its startup state and wait for reset to complete
+            d.chctrla.modify(|_, w| w.swrst().set_bit());
```

Comment:
You reset the channel a few lines above, but then you do it again here. Am I missing something? Or is that just redundant?

Reply:
You're right, that was an oversight on my part. No need to reset the channel twice.